### ISSUE TITLE
Warning 'protocol has already been consolidated' when recompiling

### ISSUE THREAD

**Issue by beno** [elixir-lang/elixir#5228]
When I add a protocol implementation such as the following, I run into a waring when recompiling:

``` elixir
defimpl Poison.Encoder, for: BSON.ObjectId do
  def encode(id, options) do
    BSON.ObjectId.encode!(id) |> Poison.Encoder.encode(options)
  end
end
```

produces on recompile

```
warning: the Poison.Encoder protocol has already been consolidated, an implementation for BSON.ObjectId has no effect
```

Using Elixir 1.3.2, for more info see https://github.com/ericmj/mongodb/issues/46#issuecomment-247893455

**Comment by ericmj** [comment]
This happens when calling `recompile/0` in iex.

**Comment by liveresume** [comment]
I believe this is happening to you on auto-recompile as it is for me.

If you make changes to the Encoder you have to stop the app and restart to see the changes.

**Comment by chaoqu** [comment]
has this regressed? I am using Elixir 1.9.0-rc.0 (aad7aa4) (compiled with Erlang/OTP 20)

**Comment by supernintendo** [comment]
This seems to have regressed. I am using Elixir 1.8.1 with Erlang/OTP 20 and get the warning any time changed files are auto-recompiled in development.

**Comment by noelia-lencina** [comment]
I'm seeing the same warning with 1.8.2 compiled for Erlang/OTP 21. I can reproduce it every time by running `r/1` in the iex console with the name of the module that's implementing the protocol (in this case the protocol is `Phoenix.HTML.Safe`).

**Comment by augnustin** [comment]
Same here, happens at recompilation with

```
elixir --version
Erlang/OTP 21 [erts-10.3.5.6] [source] [64-bit] [smp:4:4] [ds:4:4:10] [async-threads:1] [hipe]
Elixir 1.8.2 (compiled with Erlang/OTP 21)
```

Would be cleaner without it. :smile:

**Comment by KalvinHom** [comment]
Hitting this all the time as well, cannot auto-recompile at all, have to manually restart the server for any change.

Erlang/OTP 22 [erts-10.4.3] [source] [64-bit] [smp:16:16] [ds:16:16:10] [async-threads:1] [hipe]
Elixir 1.9.0 (compiled with Erlang/OTP 20)

**Comment by nathanaelreges** [comment]
Including: ` consolidate_protocols: Mix.env() != :dev` in mix.exs seems to have solved it for me.
Like suggested here: https://hexdocs.pm/elixir/Protocol.html#module-consolidation

**Comment by bcardarella** [comment]
I'm hitting this today, I cannot resolve the warning even after adding the above mentioned changed to `mix.exs`

**Comment by josevalim** [comment]
@bcardarella can you try deleting `_build` and let me know if it works after that? I think I know why the change may not be propagating after the project is compiled.

**Comment by bcardarella** [comment]
@josevalim still happening, fwiw this is in VSCode. I'm not sure if there are build artifacts that I would need to clear out otherwise.

**Comment by josevalim** [comment]
@bcardarella they use a separate folder, _elixir_ls or .elixir_ls or similar.

**Comment by dennym** [comment]
Just hijacking this issue quickly. Currently running in this issue with Poison and the usage of `defimpl`.
```
 the Poison.Decoder protocol has already been consolidated, an implementation for Foo.Bar has no effect. If you want to implement protocols after compilation or during tests, check the "Consolidation" section in the Protocol module documentation
    │
  5 │ defimpl Poison.Decoder, for: Foo.Bar do
    │ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    │
    └─ lib/foo/bar.ex:5: (file)
```

This happens when using `recompile/0` in `iex -S mix`. Usually after this I also have to hit a `mix clean`. Killing and restarting the iex does not lead to this issue.

Erlang/OTP 26
Elixir 1.16.0


This has no additonal tooling which could interfere and its on lain old zsh/bash.

**Comment by josevalim** [comment]
Thank you, I have fixed it in main and I will backport to v1.17 too.
